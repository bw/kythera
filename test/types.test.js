require("jest")
const common = require("./common")

let i = 0

afterEach(() => {
	i += 1
})

describe("Types", () => {
	let program = common.parseFile("test/vars.ky")
	i = 0

	let t = (name) => {
		test(name, () => expect(program[i]).toMatchSnapshot())
	}

	describe("'new'", () => {
		t("int")
		t("float")
		t("str")
		t("null")
	})

	describe("'name'", () => {
		t("int primitive")
		t("fn type")
		t("obj type")
	})

	describe("'let'", () => {
		t("From int literal")
		t("From 'new'")
		t("From obj literal")
	})

	describe("obj access", () => {
		t("Access")
		t("Access")
	})

	describe("Casting", () => {
		t("Declaration from type cast")
		t("Typeof")
		t("Typeof comparison")
		t("Typeof comparison with nested type cast")
		t("Function using user-defined type")
		t("Function call with type cast")
		t("Function call without type cast")
	})
})