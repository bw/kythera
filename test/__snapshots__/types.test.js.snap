// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Types 'let' From 'new' 1`] = `
Object {
  "identifier": "b",
  "kind": "let",
  "value": Object {
    "kind": "new",
    "target": Object {
      "kind": "type",
      "origin": "builtin",
      "type": "int",
    },
  },
}
`;

exports[`Types 'let' From int literal 1`] = `
Object {
  "identifier": "a",
  "kind": "let",
  "value": Object {
    "kind": "literal",
    "type": "int",
    "value": 10,
  },
}
`;

exports[`Types 'let' From obj literal 1`] = `
Object {
  "identifier": "myObj",
  "kind": "let",
  "value": Object {
    "kind": "literal",
    "type": "obj",
    "value": Object {
      "a": Object {
        "kind": "literal",
        "type": "int",
        "value": 99,
      },
      "b": Object {
        "kind": "literal",
        "type": "str",
        "value": "beep",
      },
    },
  },
}
`;

exports[`Types 'name' fn type 1`] = `
Object {
  "kind": "name",
  "name": "myFnType",
  "target": Object {
    "kind": "type",
    "origin": "builtin",
    "parameters": Array [
      Object {
        "kind": "type",
        "origin": "builtin",
        "type": "int",
      },
      Object {
        "kind": "type",
        "origin": "builtin",
        "type": "str",
      },
      Object {
        "kind": "type",
        "name": "myIntType",
        "origin": "named",
      },
    ],
    "returns": Object {
      "kind": "type",
      "origin": "builtin",
      "type": "str",
    },
    "type": "fn",
  },
}
`;

exports[`Types 'name' int primitive 1`] = `
Object {
  "kind": "name",
  "name": "myIntType",
  "target": Object {
    "kind": "type",
    "origin": "builtin",
    "type": "int",
  },
}
`;

exports[`Types 'name' obj type 1`] = `
Object {
  "kind": "name",
  "name": "myObjType",
  "target": Object {
    "kind": "type",
    "origin": "builtin",
    "structure": Object {
      "a": Object {
        "kind": "type",
        "origin": "builtin",
        "type": "int",
      },
      "b": Object {
        "kind": "type",
        "origin": "builtin",
        "type": "str",
      },
    },
    "type": "obj",
  },
}
`;

exports[`Types 'new' float 1`] = `
Object {
  "kind": "new",
  "target": Object {
    "kind": "type",
    "origin": "builtin",
    "type": "float",
  },
}
`;

exports[`Types 'new' int 1`] = `
Object {
  "kind": "new",
  "target": Object {
    "kind": "type",
    "origin": "builtin",
    "type": "int",
  },
}
`;

exports[`Types 'new' null 1`] = `
Object {
  "kind": "new",
  "target": Object {
    "kind": "type",
    "origin": "builtin",
    "type": "null",
  },
}
`;

exports[`Types 'new' str 1`] = `
Object {
  "kind": "new",
  "target": Object {
    "kind": "type",
    "origin": "builtin",
    "type": "str",
  },
}
`;

exports[`Types Casting Declaration from type cast 1`] = `
Object {
  "identifier": "myCastedObj",
  "kind": "let",
  "value": Object {
    "from": Object {
      "kind": "identifier",
      "name": "myObj",
    },
    "kind": "as",
    "to": Object {
      "kind": "type",
      "name": "myObjType",
      "origin": "named",
    },
  },
}
`;

exports[`Types Casting Function call with type cast 1`] = `
Object {
  "arguments": Array [
    Object {
      "from": Object {
        "kind": "identifier",
        "name": "myObj",
      },
      "kind": "as",
      "to": Object {
        "kind": "type",
        "name": "myObjType",
        "origin": "named",
      },
    },
  ],
  "kind": "call",
  "target": Object {
    "kind": "identifier",
    "name": "mySpecificFn",
  },
}
`;

exports[`Types Casting Function call without type cast 1`] = `
Object {
  "arguments": Array [
    Object {
      "kind": "identifier",
      "name": "myCastedObj",
    },
  ],
  "kind": "call",
  "target": Object {
    "kind": "identifier",
    "name": "mySpecificFn",
  },
}
`;

exports[`Types Casting Function using user-defined type 1`] = `
Object {
  "identifier": "mySpecificFn",
  "kind": "let",
  "value": Object {
    "body": Array [
      Object {
        "kind": "objAccess",
        "member": "a",
        "obj": "object",
      },
      Object {
        "kind": "objAccess",
        "member": "b",
        "obj": "object",
      },
      Object {
        "kind": "return",
        "value": Object {
          "kind": "literal",
          "type": "int",
          "value": 10,
        },
      },
    ],
    "kind": "literal",
    "parameters": Array [
      Object {
        "name": "object",
        "type": Object {
          "kind": "type",
          "name": "myObjType",
          "origin": "named",
        },
      },
    ],
    "returns": Object {
      "kind": "type",
      "origin": "builtin",
      "type": "null",
    },
    "type": "fn",
  },
}
`;

exports[`Types Casting Typeof 1`] = `
Object {
  "kind": "typeof",
  "target": Object {
    "kind": "identifier",
    "name": "myCastedObj",
  },
}
`;

exports[`Types Casting Typeof comparison 1`] = `
Object {
  "kind": "binary",
  "left": Object {
    "kind": "typeof",
    "target": Object {
      "kind": "identifier",
      "name": "myCastedObj",
    },
  },
  "operator": "==",
  "right": Object {
    "kind": "typeof",
    "target": Object {
      "kind": "identifier",
      "name": "myObj",
    },
  },
}
`;

exports[`Types Casting Typeof comparison with nested type cast 1`] = `
Object {
  "kind": "binary",
  "left": Object {
    "kind": "typeof",
    "target": Object {
      "kind": "identifier",
      "name": "myCastedObj",
    },
  },
  "operator": "==",
  "right": Object {
    "kind": "typeof",
    "target": Object {
      "from": Object {
        "kind": "identifier",
        "name": "myObj",
      },
      "kind": "as",
      "to": Object {
        "kind": "type",
        "name": "myObjType",
        "origin": "named",
      },
    },
  },
}
`;

exports[`Types obj access Access 1`] = `
Object {
  "kind": "objAccess",
  "member": "a",
  "obj": "myObj",
}
`;

exports[`Types obj access Access 2`] = `
Object {
  "kind": "objAccess",
  "member": "b",
  "obj": "myObj",
}
`;
